File: common/src/main/java/BallControl.java
import com.fasterxml.jackson.annotation.JsonProperty;

import java.security.SecureRandom;

public class BallControl {

    public static Rs ballKept(Point pos) {
        return new Rs(pos, true, false);
    }

    public static Rs ballTaken(Point pos) {
        return new Rs(pos, true, true);
    }

    public static Rs ballMissed(Point pos) {
        return new Rs(pos, false, false);
    }


    record Rq(
            @JsonProperty("pos") Point pos,
            @JsonProperty("player") Player player
    ) implements Jsonisable {}

    record Rs(
            @JsonProperty("pos") Point pos,
            @JsonProperty("success") boolean success,
            @JsonProperty("takesBall") boolean takesBall
    ) implements Jsonisable {}

    record Player(
            @JsonProperty("id") String id,
            @JsonProperty("skill") int skill,
            @JsonProperty("name") String name,
            @JsonProperty("team") String team
    ) implements Jsonisable {
        boolean takesBall(Player controllingPlayer, SecureRandom rnd) {
            return controllingPlayer == null || rnd.nextInt(2 * controllingPlayer.skill + skill) < skill;
        }

        public Player withSkill(int skill) {
            return new Player(id, skill, name, team);
        }
    }

    record Collision(
            @JsonProperty("pos") Point pos,
            @JsonProperty("vec") Point vec
    ) implements Jsonisable {}
}

File: common/src/main/kotlin/demo/mesharena/common/BallControl.kt
package demo.mesharena.common

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import java.security.SecureRandom


class BallControl {
  data class Rq @JsonCreator constructor(
    @JsonProperty("pos") val pos: Point,
    @JsonProperty("player") val player: Player
  ) : Jsonisable

  data class Rs @JsonCreator constructor(
    @JsonProperty("pos") val pos: Point,
    @JsonProperty("success") val success: Boolean,
    @JsonProperty("takesBall") val takesBall: Boolean
  ) : Jsonisable

  data class Player @JsonCreator constructor(
    @JsonProperty("id") val id: String,
    @JsonProperty("skill") var skill: Int,
    @JsonProperty("name") val name: String,
    @JsonProperty("team") val team: String
  ) : Jsonisable {
    fun takesBall(controllingPlayer: Player?, rnd: SecureRandom): Boolean {
      return controllingPlayer == null || rnd.nextInt(2 * controllingPlayer.skill + skill) < skill
    }
  }

  companion object {
    fun ballKept(pos: Point): Rs {
      return Rs(pos, success = true, takesBall = false)
    }

    fun ballTaken(pos: Point): Rs {
      return Rs(pos, success = true, takesBall = true)
    }

    fun ballMissed(pos: Point): Rs {
      return Rs(pos, success = false, takesBall = false)
    }
  }
}
